### ISSUE TITLE
HTTP client hangs with an infinite loop when an exception is thrown in the SSL Context

### ISSUE THREAD

**Issue by tanaka-de-silva** [http4s/blaze#75]
Hi, please can you help to debug an issue we're having when consuming an SSL service with http4s/blaze?

HTTP client hangs with an infinite loop when an exception is thrown in the SSL Context when the server presents 
an unknown certificate (expected behaviour)

```
10:58:36.420 [FixedPoolLoop-1] ERROR org.http4s.blaze.pipeline.Stage - org.http4s.blaze.pipeline.stages.QuietTimeoutStage Stage: 30 seconds received unhandled error command
javax.net.ssl.SSLException: Received fatal alert: certificate_unknown
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:208)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1666)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1634)
	at sun.security.ssl.SSLEngineImpl.recvAlert(SSLEngineImpl.java:1800)
	at sun.security.ssl.SSLEngineImpl.readRecord(SSLEngineImpl.java:1083)
	at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:907)
	at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at org.http4s.blaze.pipeline.stages.SSLStage.sslHandshakeLoop$1(SSLStage.scala:125)
	at org.http4s.blaze.pipeline.stages.SSLStage.liftedTree1$1(SSLStage.scala:167)
	at org.http4s.blaze.pipeline.stages.SSLStage.sslHandshake(SSLStage.scala:167)
	at org.http4s.blaze.pipeline.stages.SSLStage.$anonfun$sslHandshake$1(SSLStage.scala:130)
	at org.http4s.blaze.pipeline.stages.SSLStage.$anonfun$sslHandshake$1$adapted(SSLStage.scala:128)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
...
```

Stack trace from the debugger:

```
"http4s-blaze-client-2@9170" daemon prio=5 tid=0x142 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.http4s.blaze.pipeline.stages.SSLStage.sslHandshakeLoop$1(SSLStage.scala:134)
	  at org.http4s.blaze.pipeline.stages.SSLStage.liftedTree1$1(SSLStage.scala:167)
	  at org.http4s.blaze.pipeline.stages.SSLStage.sslHandshake(SSLStage.scala:167)
	  - locked <0x242e> (a scala.collection.mutable.ListBuffer)
	  at org.http4s.blaze.pipeline.stages.SSLStage.$anonfun$syncWrite$2(SSLStage.scala:258)
	  at org.http4s.blaze.pipeline.stages.SSLStage.$anonfun$syncWrite$2$adapted(SSLStage.scala:257)
	  at org.http4s.blaze.pipeline.stages.SSLStage$$Lambda$1607.843416079.apply(Unknown Source:-1)
	  at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	  at org.http4s.blaze.util.Execution$$anon$1.execute(Executor.scala:27)
	  at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)
	  at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallback(Promise.scala:312)
	  at scala.concurrent.impl.Promise$DefaultPromise.onComplete(Promise.scala:303)
	  at org.http4s.blaze.pipeline.stages.SSLStage.syncWrite(SSLStage.scala:260)
	  at org.http4s.blaze.pipeline.stages.SSLStage.writeRequest(SSLStage.scala:49)
	  at org.http4s.blaze.pipeline.stages.SSLStage.writeRequest(SSLStage.scala:21)
	  at org.http4s.blaze.pipeline.Tail.channelWrite(Stages.scala:84)
	  at org.http4s.blaze.pipeline.Tail.channelWrite$(Stages.scala:82)
	  at org.http4s.client.blaze.ReadBufferStage.channelWrite(ReadBufferStage.scala:14)
```

Is there any way we could avoid it looping infinitely and handle this error on the application side?

Thanks

**Comment by tanaka-de-silva** [comment]
more info:

![screen shot 2017-05-22 at 13 33 14](https://cloud.githubusercontent.com/assets/1861783/26309017/51e705e2-3ef3-11e7-99cf-872385b78597.png)

**Comment by rossabaker** [comment]
Which version are you on?  There was a change in this area recently.  That manifested itself as a StackOverflowError as opposed to a hang.  Is there a public server against which this is repeatable?  Thanks!

**Comment by tanaka-de-silva** [comment]
Sorry for omitting that info,  we are using 0.15.12. We have seen stack overflows before the upgrade (0.15.7).

**Comment by ChristopherDavenport** [comment]
I can confirm this error. As it is now tail-recursive it no longer overflows but will instead loop forever.

**Comment by rossabaker** [comment]
I have remained blissfully ignorant of that loop.  What progress are we expecting to make by looping and what reasonable break condition might we put on it?  I can study it later, but perhaps somebody already better understands that API.

**Comment by bryce-anderson** [comment]
I suspect the client is trying to perform another channel operation (say first it writes and then reads) and the second operation starts the loop after the first one results in the handshake failure exception. We should probably signal that the connection is borked if handshaking fails.

**Comment by svalaskevicius** [comment]
should it stop if the engine.unwrap result returns `Status = Closed`?

**Comment by bryce-anderson** [comment]
@svalaskevicius, yes, I think that is the root problem here. I've pushed a branch that appears to fix your test case (where I believe fixed means non-infinite loop, its still not doing your `println(..)` statements as I would have hoped, but that may be in the http4s-blaze package). Do you want to confirm that if fixes your specific problem before we cut a release?

**Comment by tanaka-de-silva** [comment]
@bryce-anderson Thank you very much for addressing the issue so quickly, it appears that the change has fixed the test case that was failing.

**Comment by bryce-anderson** [comment]
@tanaka-de-silva, great. We'll try and get this released sometime today.

**Comment by rossabaker** [comment]
:+1: to the commit, which I can't find a PR for.  We can do an http4s bump too, just so people get the fixed blaze out of the box.

**Comment by bryce-anderson** [comment]
I didn't put up a PR because I don't know why. If you want to cherry-pick
that commit & change the commit message since it's inaccurate, feel free to
merge it. Otherwise, I'll take care of it this evening.

**Comment by bryce-anderson** [comment]
This has been fixed in both master and 0.12.6 which is on its way to Central. Thanks again @tanaka-de-silva and @svalaskevicius!
